src/main/java/fi/aalto/cs/apluscourses/utils/PluginAutoInstaller.java

```diff
@@ -0,0 +1,151 @@
+import com.intellij.ide.plugins.PluginManagerMain;
+import com.intellij.ide.plugins.PluginNode;
+import com.intellij.openapi.application.ApplicationManager;
+import com.intellij.openapi.application.ModalityState;
+import com.intellij.openapi.application.WriteAction;
+import com.intellij.openapi.extensions.PluginId;
+import com.intellij.openapi.progress.ProgressManager;
+import com.intellij.openapi.project.Project;
+import com.intellij.util.concurrency.annotations.RequiresEdt;
+import fi.aalto.cs.apluscourses.intellij.notifications.NetworkErrorNotification;
+import fi.aalto.cs.apluscourses.intellij.notifications.Notifier;
+import fi.aalto.cs.apluscourses.ui.utils.PluginInstallerCallback;
+import java.io.IOException;
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.slf4j.Logger;
+
+public class PluginAutoInstaller {
+
+  private static final Logger logger = APlusLogger.logger;
+
+  private static boolean shouldDownloadPlugin(@NotNull PluginId id) {
+    return !PluginManagerCore.isPluginInstalled(id);
+  }
+
+  private static boolean shouldEnablePlugin(@NotNull PluginId id) {
+    final var plugin = PluginManagerCore.getPlugin(id);
+    return plugin != null && !plugin.isEnabled();
+  }
+
+  /**
+   * Verifies that all plugins with given plugin IDs are installed and enabled. If not, the function
+   * asks the user for consent to install and enable all missing plugins.
+   * @param project The project which initiated the operation.
+   * @param notifier A notifier instance, used to show a notification if something goes wrong.
+   * @param pluginNames List of plugin ID strings. The plugin ID can be found on the plugin's
+   *                    JetBrains Marketplace website.
+   * @param callback An interface implementing functions invoked by this method. The functions
+   *                 are related to showing the consent dialog box to the user.
+   * @return True, if all dependencies are satisfied and there's nothing to be done. False, if
+   * some dependencies were missing, and they were installed or enabled. Null, if the operation
+   * has been cancelled due to missing consent or an error.
+   */
+  @RequiresEdt
+  public static Boolean ensureDependenciesInstalled(@Nullable Project project,
+                                                    @Nullable Notifier notifier,
+                                                    @NotNull List<PluginDependency> pluginNames,
+                                                    @NotNull PluginInstallerCallback callback) {
+    final var pluginIDs = pluginNames.stream().map(p -> PluginId.getId(p.getId())).collect(Collectors.toList());
```

Comment:
Code style question: why are you using `final` with local variables (even when those locals are not referred from lambdas, right?) I don't say it's bad, just curious. :stuck_out_tongue:

Reply:
Okay! :+1: Being consistent makes sense. No need to change that.